### PR TITLE
Feature/importer

### PR DESCRIPTION
**Pull request by aapomm**
### Summary

Initial implementation for the dradis-pentera upload integration

### Testing Steps

1. Install the addon
2. Upload the fixture in the repo
3. Confirm that the node, issue, and evidence are created

### Other Information

If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

Thanks for contributing to Dradis!

### Copyright assignment

Collaboration is difficult with commercial closed source but we want
to keep as much of the OSS ethos as possible available to users
who want to fix it themselves.

In order to unambiguously own and sell Dradis Framework commercial
products, we must have the copyright associated with the entire
codebase. Any code you create which is merged must be owned by us.
That's not us trying to be a jerks, that's just the way it works.

You can delete this section, but the following sentence needs to
remain in the PR's description:

> I assign all rights, including copyright, to any future Dradis
> work by myself to Security Roots.

### Check List

- [x] Added a CHANGELOG entry
- [x] Added specs

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,8 +3,4 @@
 /log/*.log
 /pkg/
 /tmp/
-/test/dummy/db/*.sqlite3
-/test/dummy/db/*.sqlite3-*
-/test/dummy/log/*.log
-/test/dummy/storage/
-/test/dummy/tmp/
+/spec/internal/
```
Comment (caitmich) [inline .gitignore:6]: I don't see this anywhere? Is this line necessary?

Comment (caitmich) [inline .gitignore:8]: maybe add a comment about this being for combustion?

Also do we want to include Gemfile.lock here?

File: dradis-pentera.gemspec [hunk 2]
```diff
@@ -17,6 +17,7 @@ Gem::Specification.new do |spec|
 
   spec.add_dependency 'dradis-plugins', '~> 4.0'
```
Comment (caitmich) [inline dradis-pentera.gemspec:19]: Any reason for not including bundler and rake?
```
 spec.add_development_dependency 'bundler'
 spec.add_development_dependency 'rake'
```

File: lib/dradis/plugins/pentera/importer.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    # the dropdown list and uploads a file.
+    # @returns true if the operation was successful, false otherwise
+    def import(params = {})
+      file_content = File.read(params[:file])
+
+      # Parse the uploaded file into a Ruby Hash
+      logger.info { "Parsing Pentera output from #{ params[:file] }..." }
+      @data = MultiJson.decode(file_content)
+      logger.info { 'Done.' }
+
```
Comment (caitmich) [inline lib/dradis/plugins/pentera/importer.rb:17]: We should have some kind of check here to ensure that the file is what we expect ex. look for a specific key, or at least check that the file is `json`

File: lib/dradis/plugins/pentera/importer.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    def parse_hosts
+      @data['hosts'].each do |host|
+        host_label = host['hostname'] || host['ip']
+        logger.info { "\tHost: #{host_label}" }
+
+        node = content_service.create_node(label: host_label)
+
+        node.set_property(:ip, host['ip']) if host['ip']
+        node.set_property(:hostname, host['hostname']) if host['hostname']
+        node.set_property(:os, host['hostname']) if host['hostname']
```
Comment (caitmich) [inline lib/dradis/plugins/pentera/importer.rb]:
```suggestion
        node.set_property(:os, host['os_name']) if host['os_name']
```
(or `os_type`)?

File: lib/dradis/plugins/pentera/importer.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        node.set_property(:identifier, host['id']) if host['id']
+
+        if host['domain']
+          node.set_property(:domain_name, host['domain']['name'])
+          node.set_property(:fqdn, host['domain']['fqdn'])
```
Comment (caitmich) [inline lib/dradis/plugins/pentera/importer.rb]:
```suggestion
          node.set_property(:fqdn, host['domain']['FQDN'])
```

File: lib/dradis/plugins/pentera/importer.rb [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        })
+      end
+    end
+
+    def parse_vulnerabilities
+      @data['vulnerabilities'].each do |vulnerability|
+        issue_text = template_service.process_template(template: 'issue', data: vulnerability)
+        issue = content_service.create_issue(text: issue_text, id: vulnerability['id'])
+
+        node = @node_cache[vulnerability['target_id']]
```
Comment (caitmich) [inline lib/dradis/plugins/pentera/importer.rb:74]: In the sample file, some of the 'target' values in a vulnerability are 'Domain' rather than 'Host'. In that case, the `target_id` doesn't match up with a node ('Host'). So if the vulnerability's target is 'Domain', we won't be importing evidence for that issue. Is this intended?

File: lib/tasks/thorfile.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
+class PenteraTasks < Thor
+  include Rails.application.config.dradis.thor_helper_module
+
+  namespace 'dradis:plugins:pentera'
+
+  desc      'upload FILE', 'upload Pentera results in JSON format'
+  long_desc 'This plugin expect a JSON file generated by Pentera'
```
Comment (caitmich) [inline lib/tasks/thorfile.rb]:
```suggestion
  long_desc 'This plugin expects a JSON file generated by Pentera'
```

File: dradis-pentera.gemspec (hunk 2 above)
Comment (aapomm) [reply dradis-pentera.gemspec:19]: Do we need these? `rake build` seems to work as normal and running the rake tasks in dradis works as expected.

File: .gitignore (hunk 1 above)
Comment (aapomm) [reply .gitignore:8]: Ideally, Gemfile.lock is among the tracked files so the bundle environment will be similar among devs for situations like running the specs and building the gem.

Added comment ✅

File: lib/dradis/plugins/pentera/importer.rb (hunk 3 above)
Comment (aapomm) [reply lib/dradis/plugins/pentera/importer.rb:17]: Added an error check to see if the json is valid

File: lib/dradis/plugins/pentera/importer.rb (hunk 6 above)
Comment (aapomm) [reply lib/dradis/plugins/pentera/importer.rb:74]: Good point. I'm not sure what to make of this data. Let's wait for the user feedback to be sure.

File: Gemfile.lock [hunk 8]
```diff
@@ -0,0 +1,106 @@
+PATH
```
Comment (etdsoft) [inline Gemfile.lock]: @caitmich why is this file here?

File: lib/dradis/plugins/pentera/importer.rb [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+    # The framework will call this function if the user selects this plugin from
+    # the dropdown list and uploads a file.
+    # @returns true if the operation was successful, false otherwise
+    def import(params = {})
+      file_content = File.read(params[:file])
+
+      # Parse the uploaded file into a Ruby Hash
+      logger.info { "Parsing Pentera output from #{ params[:file] }..." }
+      @data = MultiJson.decode(file_content)
```
Comment (etdsoft) [inline lib/dradis/plugins/pentera/importer.rb:15]: is this the secure version?

File: lib/dradis/plugins/pentera/importer.rb [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+      logger.info { 'Done.' }
+
+      @node_cache = {}
+      parse_hosts
+      parse_vulnerabilities
+
+    rescue MultiJson::ParseError
+      logger.error 'ERROR: invalid JSON file uploaded. '\
+        'Are you sure you uploaded a Pentera file?'
+      exit(-1)
```
Comment (etdsoft) [inline lib/dradis/plugins/pentera/importer.rb]: This is probably not what you want

File: spec/dradis/plugins/pentera/importer_spec.rb [hunk 11]
```diff
@@ -0,0 +1,49 @@
+require 'spec_helper'
+
+describe Dradis::Plugins::Pentera::Importer do
```
Comment (etdsoft) [inline spec/dradis/plugins/pentera/importer_spec.rb:3]: we're only testing the happy path, otherwise I'd suspect you'd have picked up my previous comment.

File: dradis-pentera.gemspec [hunk 12]
```diff
@@ -17,6 +17,7 @@ Gem::Specification.new do |spec|
...
 
-  s.add_dependency 'combustion', '~> 0.6.0'
-  s.add_dependency 'rspec-rails'
+  spec.add_development_dependency 'rspec-rails'
+  spec.add_development_dependency 'combustion', '~> 0.5.2'
```
Comment (etdsoft) [inline dradis-pentera.gemspec:21]: are we using this?

File: Gemfile.lock (hunk 8 above)
Comment (aapomm) [reply Gemfile.lock]: This was mentioned here: https://github.com/dradis/dradis-pentera/pull/1#discussion_r1483194402. I decided to keep this file because of this reasoning:

> Ideally, Gemfile.lock is among the tracked files so the bundle environment will be similar among devs for situations like running the specs and building the gem.

Note that the specs here can be executed locally (no need for the Dradis base) so the bundle context will matter.

File: dradis-pentera.gemspec (hunk 12 above)
Comment (aapomm) [reply dradis-pentera.gemspec:21]: Yes both are needed for the standalone specs

File: spec/dradis/plugins/pentera/importer_spec.rb (hunk 11 above)
Comment (aapomm) [reply spec/dradis/plugins/pentera/importer_spec.rb:3]: Updated to support the failure path

File: lib/dradis/plugins/pentera/importer.rb (hunk 10 above)
Comment (aapomm) [reply lib/dradis/plugins/pentera/importer.rb]: Updated to return `false` instead

File: lib/dradis/plugins/pentera/importer.rb (hunk 9 above)
Comment (aapomm) [reply lib/dradis/plugins/pentera/importer.rb:15]: I'm using `MultiJson` for consistency with dradis-brakeman although `JSON` would also work here

File: Gemfile.lock (hunk 8 above)
Comment (etdsoft) [reply Gemfile.lock]: My understanding is that .lock shouldn't be part of gem repos, and it's never been. If we need to change directions, it has to be across the board, not in a 1-off repo.

Comment (aapomm) [reply Gemfile.lock]: Makes sense. I removed it from the repo
